binderhub/repoproviders.py

```diff
@@ -83,6 +106,135 @@ def get_build_slug(self):
         return '{user}-{repo}'.format(user='Rick', repo='Morty')
 
 
+class BasicGitRepoProvider(RepoProvider):
+    """Bare bones git repo provider.
+
+    Users must provide a spec and has to specify a resolved_ref as a query parameter.
+
+    eg:
+    /build/git/https%3A//git.example.com/ns/repo.git?resolved_ref=a01ce557f868d7060a6099bd3c2d739734dcaf15
+    """
+
+    name = Unicode("BasicGit")
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.repo = self.spec
+
+        resolved_ref = extract_string_from_arguments(self.arguments, 'resolved_ref')
+        if not resolved_ref:
+            raise ValueError("`resolved_ref` must be specified as a query parameter for the basic git provider")
+        sha1_validate(resolved_ref)
+        self.resolved_ref = resolved_ref
+
+    @gen.coroutine
+    def get_resolved_ref(self):
+        return self.resolved_ref
+
+    def get_repo_url(self):
+        return self.repo
+
+    def get_build_slug(self):
+        parsed = urllib.parse.urlparse(self.spec)
+        parts = [parsed.netloc]
+        parts.extend(parsed.path.strip('/').split('/'))
+        # escape the name and replace dashes with something else.
+        return '-'.join(p.replace('-', '_-') for p in parts)
+
+
+class GitlabRepoProvider(RepoProvider):
```

Comment:
Should this derive from `BaseGitRepoProvider`?  If not, I'm not sure that the generic GitRepoProvider class above is necessary.